### PR TITLE
Reset the VF if failure occurs before the netconf is cached

### PR DESCRIPTION
**Pull request by adilGhaffarDev**
This PR fixes following issue:
#212

### CODE REVIEW

**Comment by SchSeba** [comment]
Thanks for the PR!

/lgtm

File: cmd/sriov/main.go [hunk 1]
```diff
@@ -94,6 +94,8 @@ func cmdAdd(args *skel.CmdArgs) error {
 				})
 				if err == nil {
 					_ = sm.ReleaseVF(netConf, args.IfName, args.ContainerID, netns)
+					// Reset the VF if failure occurs before the netconf is cached
+					_ = sm.ResetVFConfig(netConf)
```
[the remark this answers is not in the document]
Comment (adilGhaffarDev) [reply cmd/sriov/main.go]: I have added the requested changes kindly check.

File: cmd/sriov/main.go [hunk 2]
```diff
@@ -74,7 +74,24 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+	defer func() {
+		if err != nil {
+			err := netns.Do(func(_ ns.NetNS) error {
+				_, err := netlink.LinkByName(args.IfName)
+				return err
+			})
+			if err == nil {
+				_ = sm.ReleaseVF(netConf, args.IfName, args.ContainerID, netns)
+				// Reset the VF if failure occurs before the netconf is cached
+				_ = sm.ResetVFConfig(netConf)
```
Comment (adrianchiris) [inline cmd/sriov/main.go:90]: we should always do this IMO, regardless of the fact that netdev is present in container namespace.
currently if ApplyVFConfig fails it will not call ResetVFConfig here.

so lets call it after L91

File: pkg/sriov/sriov.go [hunk 3]
```diff
@@ -126,7 +126,8 @@ type Manager interface {
 	SetupVF(conf *sriovtypes.NetConf, podifName string, cid string, netns ns.NetNS) (string, error)
 	ReleaseVF(conf *sriovtypes.NetConf, podifName string, cid string, netns ns.NetNS) error
 	ResetVFConfig(conf *sriovtypes.NetConf) error
-	ApplyVFConfig(conf *sriovtypes.NetConf) error
+	ApplyVFConfig(conf *sriovtypes.NetConf, pfLink netlink.Link) error
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go:129]: why did you change this interface method ?  just to avoid getting netlink handle twice ?
I think it should be kept as is for the sake of simplicity as before

File: pkg/sriov/sriov.go [hunk 4]
```diff
@@ -126,7 +126,8 @@ type Manager interface {
...
+	FillOriginalVfInfo(conf *sriovtypes.NetConf) (netlink.Link, error)
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go:130]: the name of the method and what it returns is a bit confusing.
Id prefer that it will just do one thing and fill the original VF info.

File: pkg/sriov/sriov.go [hunk 5]
```diff
@@ -374,6 +364,22 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
...
+func (s *sriovManager) FillOriginalVfInfo(conf *sriovtypes.NetConf) (netlink.Link, error) {
+	pfLink, err := s.nLink.LinkByName(conf.Master)
+	if err != nil {
+		return pfLink, fmt.Errorf("failed to lookup master %q: %v", conf.Master, err)
+	}
+	// Save current the VF state before modifying it
+	vfState := getVfInfo(pfLink, conf.VFID)
+	if vfState == nil {
+		return pfLink, fmt.Errorf("failed to find vf %d", conf.VFID)
+		//return fmt.Errorf("failed to find vf %d", conf.VFID)
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go:377]: nit: remove line

File: pkg/sriov/sriov.go [hunk 6]
```diff
@@ -374,6 +364,22 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
 	return nil
 }
 
+// Fills original VF state
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go:367]: nit add golang docs string

File: cmd/sriov/main.go [hunk 7]
```diff
@@ -74,7 +74,24 @@ func cmdAdd(args *skel.CmdArgs) error {
 	defer netns.Close()
 
 	sm := sriov.NewSriovManager()
-	if err := sm.ApplyVFConfig(netConf); err != nil {
+	pfLink, err := sm.FillOriginalVfInfo(netConf)
+	if err != nil {
+		return fmt.Errorf("failed to lookup master %q: %v", netConf.Master, err)
```
Comment (adrianchiris) [inline cmd/sriov/main.go:79]: error string should be: "failed to get original vf information: %v"

the needed info is already part of `err`

File: pkg/sriov/sriov.go [hunk 8]
```diff
@@ -374,6 +364,22 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
...
+func (s *sriovManager) FillOriginalVfInfo(conf *sriovtypes.NetConf) (netlink.Link, error) {
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go:368]: can you add unit tests for this new method ?

**Review by adrianchiris — COMMENTED** [review]
Hey there , thx for contributing !

a few comments.

once reworked it would be great if you squash to a single commit and add a descriptive commit message.

this brings #115 one step closer to completion :)

File: pkg/sriov/sriov.go (hunk 8 above)
Comment (adilGhaffarDev) [reply pkg/sriov/sriov.go]: I have made the changes that you have asked, kindly check it.

File: cmd/sriov/main.go [hunk 9]
```diff
@@ -74,6 +74,23 @@ func cmdAdd(args *skel.CmdArgs) error {
 	defer netns.Close()
 
 	sm := sriov.NewSriovManager()
+	err = sm.FillOriginalVfInfo(netConf)
+	if err != nil {
+		return fmt.Errorf("failed to get original vf information: %v", err)
+	}
+	defer func() {
+		if err != nil {
```
Comment (wizhaoredhat) [inline cmd/sriov/main.go:82]: There is a call here
		if len(newResult.IPs) == 0 {
			return errors.New("IPAM plugin returned missing IP config")
		}
Where err wouldn't be != nil. Should we also handle that as well?

Comment (adilGhaffarDev) [reply cmd/sriov/main.go:82]: Sorry, I might be misunderstanding,` if len(newResult.IPs) == 0 ` we are returning the error and it will not be equal to nil, defer function will release ipam and vf, and resetVFConfig. Is it wrong flow? Can you elaborate bit more what is the issue here?

**Comment by adilGhaffarDev** [comment]
I would like to know how can I run these PR tests myself if I want to? I have seen /retest in one of the PR, wasn't able find any other. Thanks.

File: cmd/sriov/main.go (hunk 9 above)
Comment (wizhaoredhat) [reply cmd/sriov/main.go:82]: The defer function checks that "if err != nil" and in that specific case err will equal nil since it is a length check. So should we actually do something like this?
```
		if len(newResult.IPs) == 0 {
			err = ...
                        return err
		}
```

Comment (adilGhaffarDev) [reply cmd/sriov/main.go:82]: Sorry I missed that. Now its fixed.

Comment (wizhaoredhat) [reply cmd/sriov/main.go:82]: looks good now, thanks!

**Comment by adilGhaffarDev** [comment]
@SchSeba kindly review and trigger tests here.

**Comment by cyclinder** [comment]
/cc @adrianchiris @SchSeba 

PTAL :)

**Review by adrianchiris — APPROVED** [review]
LGTM!

**Comment by SchSeba** [comment]
This will not work as expected from here

https://github.com/k8snetworkplumbingwg/sriov-cni/pull/222/files#diff-30bddd0b0a3414968a5fa92d19e50ad3168e273728fd52ba867afbc442defe62R115

```
r, err := ipam.ExecAdd(netConf.IPAM.Type, args.StdinData)
```

you can't allocate the new err variable or it will not get caught by the defer function.

**Comment by adilGhaffarDev** [comment]
> 

Fixed, there was another place with similar allocation that is also fixed. kindly check.

**Review by SchSeba — APPROVED** [review]
Thanks @adilGhaffarDev it looks good now

/lgtm
